packages/next/src/client/components/navigation.ts

```diff
@@ -140,7 +170,44 @@ export function useRouter(): AppRouterInstance {
  */
 // Client components API
 export function useParams<T extends Params = Params>(): T {
-  return useContext(PathParamsContext) as T
+  // If there are any unknown route parameters, then we should track this as
+  // a dynamic access.
+  trackParamsAccessed('useParams()')
+
+  const params = useContext(PathParamsContext)
+
+  // Replace the unknown route params with the fallback ones.
+  if (typeof window === 'undefined' && params) {
+    const unknownRouteParams = getUnknownRouteParams()
+    if (unknownRouteParams) {
+      for (const [key, value] of unknownRouteParams) {
+        params[key] = value
+      }
+    }
+  }
```

Comment:
the unknownRouteParams branch is hit when resuming a fallback prerender right? In that case won't params already have all the filled in values? Why do we need to do this mapping?

If we are in a prerender then we wouldn't have bailed out already for dynamic access reasons.

If we are doing a dynamic render then there won't be any unknownRouteParams.

In what case is this actually necessary to do?

Reply:
During SSR resume, these are still unknown.